File: packages/sdk/python/ihui_ai/modules/knowledge.py
"""知识库 / RAG / 知识图谱模块。

端点(13 个):
- GET    /v1/knowledge/health
- GET    /v1/knowledge/documents
- POST   /v1/knowledge/documents
- GET    /v1/knowledge/documents/:id
- GET    /v1/knowledge/documents/:id/chunks
- DELETE /v1/knowledge/documents/:id
- POST   /v1/knowledge/documents/batch-delete
- POST   /v1/knowledge/search
- POST   /v1/knowledge/rag-context
- POST   /v1/knowledge-graph/extract
- POST   /v1/knowledge-graph/build
- GET    /v1/knowledge-graph/data
- DELETE /v1/knowledge-graph/data
"""

from __future__ import annotations

from urllib.parse import quote

from ..async_base import AsyncBaseClient
from ..base import BaseClient
from ..types import (
    V1BatchDeleteDocumentsRequest,
    V1BatchDeleteDocumentsResponse,
    V1DocumentChunksResponse,
    V1IngestDocumentRequest,
    V1IngestDocumentResponse,
    V1KnowledgeDocumentDetail,
    V1KnowledgeDocumentsResponse,
    V1KnowledgeGraphBuildRequest,
    V1KnowledgeGraphBuildResponse,
    V1KnowledgeGraphDataResponse,
    V1KnowledgeGraphExtractRequest,
    V1KnowledgeGraphExtractResponse,
    V1KnowledgeHealthResponse,
    V1KnowledgeSearchRequest,
    V1KnowledgeSearchResponse,
    V1RagContextRequest,
    V1RagContextResponse,
)


class KnowledgeApi:
    """知识库 / RAG / 知识图谱模块(同步)。"""

    def __init__(self, client: BaseClient) -> None:
        self._client = client

    def health(self) -> V1KnowledgeHealthResponse:
        """GET /v1/knowledge/health(健康检查)。"""
        return self._client.request("GET", "/knowledge/health")

    def list_documents(self) -> V1KnowledgeDocumentsResponse:
        """GET /v1/knowledge/documents(文档列表)。"""
        return self._client.request("GET", "/knowledge/documents")

    def ingest_document(self, req: V1IngestDocumentRequest) -> V1IngestDocumentResponse:
        """POST /v1/knowledge/documents(文档入库)。"""
        return self._client.request("POST", "/knowledge/documents", req)

    def get_document(self, document_id: str) -> V1KnowledgeDocumentDetail:
        """GET /v1/knowledge/documents/:id(文档详情)。"""
        return self._client.request("GET", f"/knowledge/documents/{quote(document_id, safe='')}")

    def get_document_chunks(self, document_id: str) -> V1DocumentChunksResponse:
        """GET /v1/knowledge/documents/:id/chunks(文档分块)。"""
        return self._client.request(
            "GET", f"/knowledge/documents/{quote(document_id, safe='')}/chunks"
        )

    def delete_document(self, document_id: str) -> None:
        """DELETE /v1/knowledge/documents/:id(删除文档)。"""
        self._client.request("DELETE", f"/knowledge/documents/{quote(document_id, safe='')}")

    def batch_delete_documents(self, req: V1BatchDeleteDocumentsRequest) -> V1BatchDeleteDocumentsResponse:
        """POST /v1/knowledge/documents/batch-delete(批量删除)。"""
        return self._client.request("POST", "/knowledge/documents/batch-delete", req)

    def search(self, req: V1KnowledgeSearchRequest) -> V1KnowledgeSearchResponse:
        """POST /v1/knowledge/search(语义搜索)。"""
        return self._client.request("POST", "/knowledge/search", req)

    def rag_context(self, req: V1RagContextRequest) -> V1RagContextResponse:
        """POST /v1/knowledge/rag-context(RAG 上下文检索)。"""
        return self._client.request("POST", "/knowledge/rag-context", req)

    def extract_graph(self, req: V1KnowledgeGraphExtractRequest) -> V1KnowledgeGraphExtractResponse:
        """POST /v1/knowledge-graph/extract(知识图谱抽取)。"""
        return self._client.request("POST", "/knowledge-graph/extract", req)

    def build_graph(self, req: V1KnowledgeGraphBuildRequest) -> V1KnowledgeGraphBuildResponse:
        """POST /v1/knowledge-graph/build(知识图谱构建)。"""
        return self._client.request("POST", "/knowledge-graph/build", req)

    def get_graph_data(self) -> V1KnowledgeGraphDataResponse:
        """GET /v1/knowledge-graph/data(知识图谱数据)。"""
        return self._client.request("GET", "/knowledge-graph/data")

    def clear_graph(self) -> None:
        """DELETE /v1/knowledge-graph/data(清空知识图谱)。"""
        self._client.request("DELETE", "/knowledge-graph/data")


class AsyncKnowledgeApi:
    """知识库 / RAG / 知识图谱模块(asyncio)。"""

    def __init__(self, client: AsyncBaseClient) -> None:
        self._client = client

    async def health(self) -> V1KnowledgeHealthResponse:
        """GET /v1/knowledge/health(健康检查)。"""
        return await self._client.request("GET", "/knowledge/health")

    async def list_documents(self) -> V1KnowledgeDocumentsResponse:
        """GET /v1/knowledge/documents(文档列表)。"""
        return await self._client.request("GET", "/knowledge/documents")

    async def ingest_document(self, req: V1IngestDocumentRequest) -> V1IngestDocumentResponse:
        """POST /v1/knowledge/documents(文档入库)。"""
        return await self._client.request("POST", "/knowledge/documents", req)

    async def get_document(self, document_id: str) -> V1KnowledgeDocumentDetail:
        """GET /v1/knowledge/documents/:id(文档详情)。"""
        return await self._client.request("GET", f"/knowledge/documents/{quote(document_id, safe='')}")

    async def get_document_chunks(self, document_id: str) -> V1DocumentChunksResponse:
        """GET /v1/knowledge/documents/:id/chunks(文档分块)。"""
        return await self._client.request(
            "GET", f"/knowledge/documents/{quote(document_id, safe='')}/chunks"
        )

    async def delete_document(self, document_id: str) -> None:
        """DELETE /v1/knowledge/documents/:id(删除文档)。"""
        await self._client.request("DELETE", f"/knowledge/documents/{quote(document_id, safe='')}")

    async def batch_delete_documents(self, req: V1BatchDeleteDocumentsRequest) -> V1BatchDeleteDocumentsResponse:
        """POST /v1/knowledge/documents/batch-delete(批量删除)。"""
        return await self._client.request("POST", "/knowledge/documents/batch-delete", req)

    async def search(self, req: V1KnowledgeSearchRequest) -> V1KnowledgeSearchResponse:
        """POST /v1/knowledge/search(语义搜索)。"""
        return await self._client.request("POST", "/knowledge/search", req)

    async def rag_context(self, req: V1RagContextRequest) -> V1RagContextResponse:
        """POST /v1/knowledge/rag-context(RAG 上下文检索)。"""
        return await self._client.request("POST", "/knowledge/rag-context", req)

    async def extract_graph(self, req: V1KnowledgeGraphExtractRequest) -> V1KnowledgeGraphExtractResponse:
        """POST /v1/knowledge-graph/extract(知识图谱抽取)。"""
        return await self._client.request("POST", "/knowledge-graph/extract", req)

    async def build_graph(self, req: V1KnowledgeGraphBuildRequest) -> V1KnowledgeGraphBuildResponse:
        """POST /v1/knowledge-graph/build(知识图谱构建)。"""
        return await self._client.request("POST", "/knowledge-graph/build", req)

    async def get_graph_data(self) -> V1KnowledgeGraphDataResponse:
        """GET /v1/knowledge-graph/data(知识图谱数据)。"""
        return await self._client.request("GET", "/knowledge-graph/data")

    async def clear_graph(self) -> None:
        """DELETE /v1/knowledge-graph/data(清空知识图谱)。"""
        await self._client.request("DELETE", "/knowledge-graph/data")


__all__ = ["KnowledgeApi", "AsyncKnowledgeApi"]

File: sdks/typescript/src/knowledge.ts
/**
 * 知识库 / RAG / 知识图谱模块。
 *
 * 端点(13 个):
 * - GET    /v1/knowledge/health
 * - GET    /v1/knowledge/documents
 * - POST   /v1/knowledge/documents
 * - GET    /v1/knowledge/documents/:id
 * - GET    /v1/knowledge/documents/:id/chunks
 * - DELETE /v1/knowledge/documents/:id
 * - POST   /v1/knowledge/documents/batch-delete
 * - POST   /v1/knowledge/search
 * - POST   /v1/knowledge/rag-context
 * - POST   /v1/knowledge-graph/extract
 * - POST   /v1/knowledge-graph/build
 * - GET    /v1/knowledge-graph/data
 * - DELETE /v1/knowledge-graph/data
 */

import type { BaseClient } from './base.js'

// =============================================================================
// 内联类型定义
// =============================================================================

/** 知识库健康检查响应。 */
export interface V1KnowledgeHealthResponse {
  status: 'ok' | 'degraded'
  documents: number
  chunks: number
}

export interface V1KnowledgeDocumentsResponse {
  object: 'list'
  data: Array<{
    id: string
    title: string
    source: string
    chunkCount: number
    sizeBytes: number
    createdAt: string
  }>
}

export interface V1IngestDocumentRequest {
  title: string
  content: string
  source?: string
  chunkStrategy?: string
  chunkSize?: number
  chunkOverlap?: number
}

export interface V1IngestDocumentResponse {
  documentId: string
  chunkCount: number
  status: 'ingested'
}

/** 文档详情。 */
export interface V1KnowledgeDocumentDetail {
  id: string
  title: string
  source: string
  chunkCount: number
  sizeBytes: number
  createdAt: string
  updatedAt: string
}

export interface V1DocumentChunksResponse {
  object: 'list'
  data: Array<{
    id: string
    content: string
    index: number
    metadata?: Record<string, unknown>
  }>
}

export interface V1KnowledgeSearchRequest {
  query: string
  topK?: number
  documentIds?: string[]
  threshold?: number
}

export interface V1KnowledgeSearchResponse {
  object: 'list'
  data: Array<{
    id: string
    documentId: string
    content: string
    score: number
    metadata?: Record<string, unknown>
  }>
}

export interface V1RagContextRequest {
  query: string
  topK?: number
  injectSystemPrompt?: boolean
}

export interface V1RagContextResponse {
  context: string
  sources: Array<{
    documentId: string
    chunkId: string
    score: number
  }>
}

export interface V1KnowledgeGraphExtractRequest {
  text: string
  extractType?: string
}

export interface V1KnowledgeGraphExtractResponse {
  entities: Array<{
    id: string
    name: string
    type: string
    properties?: Record<string, unknown>
  }>
  relations: Array<{
    source: string
    target: string
    type: string
    properties?: Record<string, unknown>
  }>
}

/** 知识图谱构建请求。 */
export interface V1KnowledgeGraphBuildRequest {
  source: string
  sourceType?: string
}

/** 知识图谱构建响应。 */
export interface V1KnowledgeGraphBuildResponse {
  graphId: string
  nodes: number
  edges: number
}

export interface V1KnowledgeGraphDataResponse {
  nodes: Array<{ id: string; label: string; type: string }>
  edges: Array<{ source: string; target: string; label: string }>
}

/** 批量删除请求。 */
export interface V1BatchDeleteDocumentsRequest {
  documentIds: string[]
}

/** 批量删除响应。 */
export interface V1BatchDeleteDocumentsResponse {
  deleted: number
}

// =============================================================================
// Module 接口 + 工厂函数
// =============================================================================

export interface KnowledgeModule {
  /** GET /v1/knowledge/health(健康检查)。 */
  health(): Promise<V1KnowledgeHealthResponse>
  /** GET /v1/knowledge/documents(文档列表)。 */
  listDocuments(): Promise<V1KnowledgeDocumentsResponse>
  /** POST /v1/knowledge/documents(文档入库)。 */
  ingestDocument(req: V1IngestDocumentRequest): Promise<V1IngestDocumentResponse>
  /** GET /v1/knowledge/documents/:id(文档详情)。 */
  getDocument(id: string): Promise<V1KnowledgeDocumentDetail>
  /** GET /v1/knowledge/documents/:id/chunks(文档分块)。 */
  getDocumentChunks(id: string): Promise<V1DocumentChunksResponse>
  /** DELETE /v1/knowledge/documents/:id(删除文档)。 */
  deleteDocument(id: string): Promise<void>
  /** POST /v1/knowledge/documents/batch-delete(批量删除)。 */
  batchDeleteDocuments(req: V1BatchDeleteDocumentsRequest): Promise<V1BatchDeleteDocumentsResponse>
  /** POST /v1/knowledge/search(语义搜索)。 */
  search(req: V1KnowledgeSearchRequest): Promise<V1KnowledgeSearchResponse>
  /** POST /v1/knowledge/rag-context(RAG 上下文检索)。 */
  ragContext(req: V1RagContextRequest): Promise<V1RagContextResponse>
  /** POST /v1/knowledge-graph/extract(知识图谱抽取)。 */
  extractGraph(req: V1KnowledgeGraphExtractRequest): Promise<V1KnowledgeGraphExtractResponse>
  /** POST /v1/knowledge-graph/build(知识图谱构建)。 */
  buildGraph(req: V1KnowledgeGraphBuildRequest): Promise<V1KnowledgeGraphBuildResponse>
  /** GET /v1/knowledge-graph/data(知识图谱数据)。 */
  getGraphData(): Promise<V1KnowledgeGraphDataResponse>
  /** DELETE /v1/knowledge-graph/data(清空知识图谱)。 */
  clearGraph(): Promise<void>
}

export function createKnowledgeModule(client: BaseClient): KnowledgeModule {
  return {
    health: () => client.request<V1KnowledgeHealthResponse>('GET', '/knowledge/health'),
    listDocuments: () =>
      client.request<V1KnowledgeDocumentsResponse>('GET', '/knowledge/documents'),
    ingestDocument: (req) =>
      client.request<V1IngestDocumentResponse>('POST', '/knowledge/documents', req),
    getDocument: (id) =>
      client.request<V1KnowledgeDocumentDetail>(
        'GET',
        `/knowledge/documents/${encodeURIComponent(id)}`,
      ),
    getDocumentChunks: (id) =>
      client.request<V1DocumentChunksResponse>(
        'GET',
        `/knowledge/documents/${encodeURIComponent(id)}/chunks`,
      ),
    deleteDocument: (id) =>
      client.request<void>('DELETE', `/knowledge/documents/${encodeURIComponent(id)}`),
    batchDeleteDocuments: (req) =>
      client.request<V1BatchDeleteDocumentsResponse>(
        'POST',
        '/knowledge/documents/batch-delete',
        req,
      ),
    search: (req) => client.request<V1KnowledgeSearchResponse>('POST', '/knowledge/search', req),
    ragContext: (req) =>
      client.request<V1RagContextResponse>('POST', '/knowledge/rag-context', req),
    extractGraph: (req) =>
      client.request<V1KnowledgeGraphExtractResponse>('POST', '/knowledge-graph/extract', req),
    buildGraph: (req) =>
      client.request<V1KnowledgeGraphBuildResponse>('POST', '/knowledge-graph/build', req),
    getGraphData: () =>
      client.request<V1KnowledgeGraphDataResponse>('GET', '/knowledge-graph/data'),
    clearGraph: () => client.request<void>('DELETE', '/knowledge-graph/data'),
  }
}
